src/optimize.jl

```diff
@@ -150,42 +151,131 @@ function optimize!(code::CodeInfo, mod::Module)
             arg1 = stmt.args[1]
-            if arg1 == :llvmcall || lookup_stmt(code.code, arg1) == Base.llvmcall
+            if (arg1 == :llvmcall || lookup_stmt(code.code, arg1) == Base.llvmcall) && isempty(sparams) && scope isa Method
                 uuid = uuid4()
                 ustr = replace(string(uuid), '-'=>'_')
                 methname = Symbol("llvmcall_", ustr)
                 nargs = length(stmt.args)-4
-                argnames = [Symbol("arg", string(i)) for i = 1:nargs]
-                # Run a mini-interpreter to extract the types
-                framecode = FrameCode(CompiledCalls, code; optimize=false)
-                frame = Frame(framecode, prepare_framedata(framecode, []))
-                idxstart = idx
-                for i = 2:4
-                    idxstart = smallest_ref(code.code, stmt.args[i], idxstart)
-                end
-                frame.pc = idxstart
-                while true
-                    pc = step_expr!(Compiled(), frame)
-                    pc == idx && break
-                    pc === nothing && error("this should never happen")
-                end
-                str, RetType, ArgType = @lookup(frame, stmt.args[2]), @lookup(frame, stmt.args[3]), @lookup(frame, stmt.args[4])
-                def = quote
-                    function $methname($(argnames...))
-                        return Base.llvmcall($str, $RetType, $ArgType, $(argnames...))
-                    end
-                end
-                f = Core.eval(CompiledCalls, def)
-                stmt.args[1] = QuoteNode(f)
-                deleteat!(stmt.args, 2:4)
+                build_compiled_call!(stmt, methname, Base.llvmcall, stmt.args[2:4], code, idx, nargs, sparams)
                 methodtables[idx] = Compiled()
             end
+        elseif isexpr(stmt, :foreigncall) && scope isa Method
+            f = lookup_stmt(code.code, stmt.args[1])
+            if isa(f, Ptr)
+                f = string(uuid4())
+            elseif isexpr(f, :call)
+                length(f.args) == 3 || continue
+                f.args[1] === tuple || continue
+                lib = f.args[3] isa String ? f.args[3] : f.args[3].value
+                prefix = f.args[2] isa String ? f.args[2] : f.args[2].value
+                f = Symbol(prefix, '_', lib)
+            end
+            # Punt on non literal ccall arguments for now
+            if !(isa(f, String) || isa(f, Symbol) || isa(f, Ptr))
+                continue
+            end
+            # TODO: Only compile one ccall per call and argument types
+            uuid = uuid4()
+            ustr = replace(string(uuid), '-'=>'_')
+            methname = Symbol("ccall", '_', f, '_', ustr)
+            nargs = stmt.args[5]
+            build_compiled_call!(stmt, methname, :ccall, stmt.args[1:3], code, idx, nargs, sparams)
+            methodtables[idx] = Compiled()
         end
     end
 
     return code, methodtables
 end
+
+function parametric_type_to_expr(t::Type)
```

Comment:
I'd guess you should put `@nospecialize(t::Type)` and haul out the anonymous function on `tv` and use similar treatment.